### PR TITLE
Import proposal block early

### PR DESCRIPTION
**Pull request by majecty**
This PR depends on #891, #933 

If a node would import a proposal block early, the node could propagate the previous block shortly, because the previous block became the best block when the proposal block is imported.

### CODE REVIEW

**Comment by majecty** [comment]
@sgkim126 Could you review this PR?
You only need to check last two commits.

**Comment by sgkim126** [comment]
Don't we need to call `mark_as_good`?

File: core/src/consensus/tendermint/mod.rs [hunk 1]
```diff
@@ -141,6 +140,34 @@ impl Tendermint {
...
+        if prev_score < new_score {
+            ctrace!(ENGINE, "future proposal changed");
+            *self.future_proposal.write() = Some(future_proposal);
+        } else {
+            ctrace!(ENGINE, "future proposal not changed");
+        }
+    }
+
+    pub fn future_proposal_prev_hash(&self) -> Option<H256> {
+        if let Some(future_proposal) = self.future_proposal.try_read_for(std::time::Duration::new(1, 0)) {
```
Comment (sgkim126) [inline core/src/consensus/tendermint/mod.rs]: Why do you use `try_read_for`?
It should be used only when there is a fallback when it failed to acquire a lock.

File: core/src/consensus/tendermint/mod.rs [hunk 2]
```diff
@@ -279,14 +293,38 @@ impl Tendermint {
...
+                        *self.proposal.write() = Some(*hash);
+                        *self.proposal_parent.write() = header.parent_hash();
+                    } else {
+                        ctrace!(ENGINE, "Proposal is received but not imported");
+                        // Proposal is received but is not verified yet.
+                        // Wait for verification.
+                        return
+                    }
+                } else {
+                    self.update_sealing()
```
Comment (sgkim126) [inline core/src/consensus/tendermint/mod.rs]: When does `get_block_hashes` return None?

**Comment by majecty** [comment]
@sgkim126 We don't need to call `mark_as_good`.
`mark_as_good` is used to remove the block/header from the verification queue.
If we import the block/header, it is removed from the verification queue while importing it.

File: core/src/consensus/tendermint/mod.rs (hunk 1 above)
Comment (majecty) [reply core/src/consensus/tendermint/mod.rs]: I'll fix it.

File: core/src/consensus/tendermint/mod.rs (hunk 2 above)
Comment (majecty) [reply core/src/consensus/tendermint/mod.rs]: `self.votes.get_block_hashes(&vote_step).first()` is true when a node received proposal message before it entered to the "propose" step.
If a node didn't receive the proposal before enter the propose step, the value is None.

**Comment by majecty** [comment]
@sgkim126 I fixed them.

File: core/src/consensus/tendermint/mod.rs (hunk 2 above)
Comment (sgkim126) [reply core/src/consensus/tendermint/mod.rs]: Is it okay to call `update_sealing`?
Even if the node is not a proposer?
Or is it guaranteed that the node is a proposer when there is no received proposal?

**Comment by sgkim126** [comment]
> mark_as_good

If then, why does `mark_as_good` exist?
Is there a case that a block should not be imported even it is good?

**Comment by majecty** [comment]
Before this commit, the proposal block is not imported although verification was successful. Then it was needed to remove the proposal from the verification queue.

Now there is no case that a block is not imported although verification was successful.

File: core/src/consensus/tendermint/mod.rs (hunk 2 above)
Comment (majecty) [reply core/src/consensus/tendermint/mod.rs]: It has been like this before.

I agree with that we do not need to call update_sealing if the node is not a proposer.
Fixing this behavior should be another PR.

**Pull request merged**